packages/gatsby/cache-dir/loader.js

```diff
@@ -1,166 +1,152 @@
-import pageFinderFactory from "./find-page"
 import emitter from "./emitter"
 import prefetchHelper from "./prefetch"
+import { match } from "@reach/router/lib/utils"
+import stripPrefix from "./strip-prefix"
+// Generated during bootstrap
+import matchPaths from "./match-paths.json"
 
 const preferDefault = m => (m && m.default) || m
 
-let devGetPageData
-let inInitialRender = true
-let hasFetched = Object.create(null)
+const pageNotFoundPaths = new Set()
+
+let apiRunner
 let syncRequires = {}
 let asyncRequires = {}
-let jsonDataPaths = {}
-let fetchHistory = []
-let fetchingPageResourceMapPromise = null
-let fetchedPageResourceMap = false
-/**
- * Indicate if pages manifest is loaded
- *  - in production it is split to separate "pages-manifest" chunk that need to be lazy loaded,
- *  - in development it is part of single "common" chunk and is available from the start.
- */
-let hasPageResourceMap = process.env.NODE_ENV !== `production`
-let apiRunner
-const failedPaths = {}
-const MAX_HISTORY = 5
 
-const jsonPromiseStore = {}
+const fetchedPageData = {}
+const pageDatas = {}
```

Comment:
Add to cache the timestamp of when a resource was fetched — after five minutes, drop item from cache and force server revalidation — that way we know we're loading fresh data.

Reply:
I really like this, but it probably deserves a bit more thought on how it impacts the rest of this file. Given that the page-data changes already address some of the issues of stale pages, I might create an issue to implement this in a future PR.

Reply:
https://github.com/gatsbyjs/gatsby/issues/13430